source/codegen/metadata_validation.py

```diff
@@ -0,0 +1,287 @@
+        # This should be DOCUMENTATION_SCHEMA but don't want to validate if codegen_method is no
+        Optional('documentation'): dict,
+        Optional('method_templates'): list,
+        Optional('custom_close'): str,
+        Optional('custom_close_method'): bool,
+        Optional("python_name"): str,
+    }
+)
+
+ATTRIBUTE_SCHEMA = Schema(
+    {
+        "name": str,
+        Optional("access"): And(str, lambda s: s in ['read', 'read only', 'write', 'write only', 'read-write']),
+        "type": str,
+        Optional("resettable"): bool,
+        Optional("enum"): str,
+        Optional("channel_based"): bool,
+        Optional("attribute_class"): str,
+        Optional("type_in_documentation"): str,
+        Optional("documentation"): DOCUMENTATION_SCHEMA,
+        Optional("lv_property"): str,
+        Optional("repeated_capability_type"): str,
+        Optional("python_type"): str,
+        Optional("python_name"): str,
+        Optional("codegen_method"): str,
+    }
+)
+
+SIMPLE_ATTRIBUTE_SCHEMA = Schema(
+    {
+        "name": str
+    }
+)
+
+ENUM_SCHEMA = Schema(
+    {
+        "values": [{
+            "name": str,
+            "value": Or(str, int, float),
+            Optional("python_name"): str,
+            Optional("documentation"): DOCUMENTATION_SCHEMA,
+        }],
+        Optional("generate-mappings"): bool,
+    }
+)
+
+def rule_is_suppressed(metadata: dict, rule: str, path: List[str]) -> bool:
+    suppression_dict_name = path[0] + "_validation_suppressions"
+    suppression_dict = metadata.get(suppression_dict_name, {})
+    for entry in path[1:-1]:
+        suppression_dict = suppression_dict.get(entry, {})
+    last_entry = path[-1]
+    if last_entry not in suppression_dict:
+        return False
+    return rule in suppression_dict[last_entry]
+
+def parameter_name_exists(function: dict, name: str) -> bool:
+    return any([param for param in function['parameters'] if param['name'] == name])
+
+def validate_function(function_name: str, metadata: dict):
+    try:
+        function: Dict[str, Any] = metadata['functions'][function_name]
+        FUNCTION_SCHEMA.validate(function)
+        if function.get('codegen_method', 'public') != 'no':
+            if 'documentation' in function:
+                DOCUMENTATION_SCHEMA.validate(function['documentation'])
+            for parameter in function['parameters']:
+                validate_parameter_size(parameter, function_name, metadata)
+                if 'type' not in parameter:
+                    if 'grpc_type' not in parameter:
+                        raise Exception(f"parameter {parameter['name']} has no type or grpc_type!")
+                    if not parameter.get('repeated_var_args', False):
+                        raise Exception(f"parameter {parameter['name']} has no type and repeated_var_args is not set!")
+                if 'enum' in parameter:
+                    if parameter['enum'] not in metadata['enums']:
```

Comment:
Is it possible to mutate the schema or have it be dynamic in a way that catches this?
(not sure that's a good idea)

Reply:
(I assume you're talking about the `repeated_var_args` part of this)

I can't think of a way to do this that would be an improvement over what we have. We could have two schemas - one where `type` is required and `repeated_var_args` is not present or False, and one where type is not present and `repeated_var_args` is True. That way everything would validate under one of those schemas. But these would have to be two additional schemas unless we wanted to copy all the other parts of the original schema into the second schema, which is bad. And I think in this case it's clearer to just do the check in code rather than adding two additional schemas for this purpose.

Reply:
Oh, actually no. I was thinking of the list of enums. Like 

`Optiona('enum'): And(str, lambda s: s in _get_actual_list_of_enums())`

Curious if that works.

I definitely don't want to worry more about repeated_var_args!